kong/runloop/wasm.lua

```diff
@@ -1,141 +1,537 @@
+---@return string?     error
+local function init_c_plan(filters)
+  if not filters then
+    return
   end
 
-  local c_ops, err = proxy_wasm.new(filters)
-  if not c_ops then
-    return nil, "proxy wasm module instantiation failed: " .. tostring(err)
+  local c_plan, err = proxy_wasm.new(filters)
+  if not c_plan then
+    return nil, "failed instantiating filter chain: "
+                .. tostring(err)
   end
 
   local ok
-  ok, err = proxy_wasm.load(c_ops)
+  ok, err = proxy_wasm.load(c_plan)
   if not ok then
-    return nil, "initial loading of proxy wasm modules failed: " .. tostring(err)
+    return nil, "failed loading filters: " .. tostring(err)
   end
 
-  return true
+  return c_plan
 end
 
 
-function _M.attach_filter_chains(ctx)
-  if not enabled then
-    return
+-- Helper method for retrieving a filter chain reference from
+-- the state table.
+--
+---@param state       kong.runloop.wasm.state
+---@param typ         integer
+---@param service_id? string
+---@param route_id?   string
+---
+---@return kong.runloop.wasm.filter_chain_reference? ref
+local function get_chain_ref(state, typ, service_id, route_id)
+  local ref
+
+  if typ == TYPE_SERVICE and service_id then
+    ref = state.by_service[service_id]
+
+  elseif typ == TYPE_ROUTE and route_id then
+    ref = state.by_route[route_id]
+
+  elseif typ == TYPE_COMBINED and service_id and route_id then
+    local routes = state.combined[service_id]
+    ref = routes and routes[route_id]
   end
 
-  local service = ctx.service or NONE
-  local route = ctx.route or NONE
+  return ref
+end
+
 
-  local filters = {}
+---
+-- Helper method for storing a new filter chain reference within
+-- the state table.
+--
+---@param state       kong.runloop.wasm.state
+---@param ref         kong.runloop.wasm.filter_chain_reference
+---@param typ         integer
+---@param service_id? string
+---@param route_id?   string
+local function store_chain_ref(state, ref, typ, service_id, route_id)
+  if typ == TYPE_SERVICE and service_id then
+    state.by_service[service_id] = ref
+
+  elseif typ == TYPE_ROUTE and route_id then
+    state.by_route[route_id] = ref
+
+  elseif typ == TYPE_COMBINED and service_id and route_id then
+    local routes = state.combined[service_id]
+
+    if not routes then
+      routes = {}
+      state.combined[service_id] = routes
+    end
 
-  for chain, err in kong.db.wasm_filter_chains:each() do
+    routes[route_id] = ref
+  end
+end
+
+
+---
+-- Build a combined filter list from 1-2 filter chain entities.
+--
+-- Disabled filter chains are skipped, and disabled filters are
+-- skipped.
+--
+-- Returns `nil` if no enabled filters are found.
+--
+---@param service_chain? kong.db.schema.entities.wasm_filter_chain
+---@param route_chain?   kong.db.schema.entities.wasm_filter_chain
+---
+---@return kong.db.schema.entities.wasm_filter[]?
+local function build_filter_list(service_chain, route_chain)
+  ---@type kong.db.schema.entities.wasm_filter[]|nil
+  local combined
```

Comment:
```suggestion
  local combined = {}
```

Doesn't that get rid of two lines? Or are you optimizing for returning nil?

Reply:
> Doesn't that get rid of two lines? Or are you optimizing for returning nil?

The latter. It's a tiny optimization, and allowing the object to be nil allows us to use nice truth-y/false-y logic elsewhere in the file.